### PR TITLE
Testing guide improvements

### PR DESCRIPTION
**Pull request by Ullaakut**
## Goal of this PR

This PR adds multiple new sections and examples to the Go testing guide.

* [Add specification for multiple testing functions on same method](https://github.com/optakt/optakt.github.io/commit/e5d1539d7258df10e2aa5432f6bbcb110eb30840)

* [Add admonition about external projects](https://github.com/optakt/optakt.github.io/commit/074292a877b078ef4c3d18e61d8c4247733e1d8c)

* [Add example for testing paths & subtest naming](https://github.com/optakt/optakt.github.io/commit/b9d0973f6c88ba28c5c492bbb54ed1daa1ae80db)

* [Specify test structure](https://github.com/optakt/optakt.github.io/commit/5c9720600655cf56da3014d50f3c6d2f82099c83)

* [Add benchmark example](https://github.com/optakt/optakt.github.io/commit/b15a320541d039b4bbdec993659f7c78d651389f)

* [Add section about testing/mocks package](https://github.com/optakt/optakt.github.io/commit/73d43bb1aa666ac80b78c065acb4e1196e57b292)

### CODE REVIEW

File: content/code/testing.md [hunk 1]
```diff
@@ -66,6 +107,11 @@ Here is an example of such a value being generated.
 
     While randomly generating valid inputs makes sense, randomly generating invalid inputs does not. In the case of invalid inputs, it is much better to have an exhaustive list of all types of cases that are expected to be invalid and always test each one of them.
 
+#### Generic Values Package
+
+When the same generic value need to be used within multiple packages, it can make sense to create a package to store common generic values.
+This package should be located at `./testing/mocks` from the root of the repository, and values should be in a file called `generic.go`.
```
Comment (AnNiran) [inline content/code/testing.md:113]: Types, values and functions/methods used for testing across the whole project can increase drastically if it is covered with tests.
Seems to me that a single file `generic.go` will be too big for this and it will contain too much different logic.
Since we have a separate sub-package for testing, it makes sense testing entities to be separated in more than one file on multiple levels; for example:
 - helpers.go - file that contains helper testing functions
 - xxx_mocks.go, xxy_mocks.go, xyy_mocks.go, etc - files that contains mocks for the different entities and their methods; can use interface for same data generation methods for example.

File: content/code/testing.md [hunk 2]
```diff
@@ -104,6 +150,20 @@ This can be very useful, as it enables creating a hierarchical structure within
...
+In those tests, usually the structure is composed of:
+
+* At the top-level, the common testing variables that are used in multiple subtests.
+    * This can be a logger that discards logs, some test input that can be reused, expected return values shared between tests, and so on.
+* In each sub-test:
+    * The test is defined as parallelized;
+    * The inputs and outputs are defined in variables;
+    * The mocks are defined and their methods overridden when necessary;
+    * The test subject is created and given the mocks;
+    * The tested method is called on the subject, its return values are often stored in `got` and `err`: `got, err := mything.Do(input)`;
```
Comment (AnNiran) [inline content/code/testing.md:164]: What is received from the test call is more like an expected value, that we also want to receive from the function call that is tested and compared against.
I would replace the word `got` with `expected`

File: content/code/testing.md [hunk 3]
```diff
@@ -66,6 +107,11 @@ Here is an example of such a value being generated.
 
     While randomly generating valid inputs makes sense, randomly generating invalid inputs does not. In the case of invalid inputs, it is much better to have an exhaustive list of all types of cases that are expected to be invalid and always test each one of them.
 
+#### Generic Values Package
```
[the remark this answers is not in the document]
Comment (Ullaakut) [reply content/code/testing.md:110]: The only types defined in this package are for mocks, so that's what the other section is about. When it comes to generic values, we do not need to expose new types since we're just trying to generate values for the existing types that need to be mocked. Does that make sense?

File: content/code/testing.md (hunk 1 above)
Comment (Ullaakut) [reply content/code/testing.md:113]: Currently the Flow DPS is our largest project and it's still in a state where we don't really feel the need to split the generic values between multiple files.

The reason is simple: they are generic values. We do not care how they are generated, we only care about them being random in a deterministic way. Because we do not care about the implementation, it doesn't really make sense to split it into multiple files. In the case of the Flow DPS that would mean having 10 or so files, which we will probably never modify/read (source: https://github.com/optakt/flow-dps/blob/master/testing/mocks/generic.go)

IMO if we ever reach a situation where this part of the guide no longer holds true, we can modify it and deal with it accordingly, but in my personal opinion that will most likely never be the case :)

File: content/code/testing.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (AnNiran) [reply content/code/testing.md:110]: Yes, okay

File: content/code/testing.md (hunk 2 above)
Comment (Ullaakut) [reply content/code/testing.md:164]: No, this is about the actual call which is being tested.

What you are talking about refers to what you do in your test:

```go
expected := myMock.Retrieve()

got := subject.Retrieve()

assert.EqualElements(t, expected, got)
```

Which is an edge case that only happens because you test that the subject returns whatever the mock returns without any modification.

In my opinion it's not the best practice either, I would personally rather have a hardcoded `want` value and make the mock return a hardcoded value as well, because this test relies on the mock to get its output variable which is very uncommon and reduces clarity IMO.

I'm fine with it in the context of your test as long as it remains localized and does not become a habit/shortcut, but it should definitely not be in the docs IMO.

WDYT?

File: content/code/testing.md [hunk 4]
```diff
@@ -66,6 +107,11 @@ Here is an example of such a value being generated.
...
+
+When the same generic value need to be used within multiple packages, it can make sense to create a package to store common generic values.
```
[the remark this answers is not in the document]
Comment (Ullaakut) [reply content/code/testing.md:112]: See https://github.com/optakt/optakt.github.io/pull/30#discussion_r836396587

File: content/code/testing.md (hunk 2 above)
Comment (AnNiran) [reply content/code/testing.md:164]: Yes, I saw that this is about the tested function call.

File: content/code/testing.md [hunk 5]
```diff
@@ -104,6 +150,20 @@ This can be very useful, as it enables creating a hierarchical structure within
...
+
+In those tests, usually the structure is composed of:
+
+* At the top-level, the common testing variables that are used in multiple subtests.
+    * This can be a logger that discards logs, some test input that can be reused, expected return values shared between tests, and so on.
+* In each sub-test:
+    * The test is defined as parallelized;
+    * The inputs and outputs are defined in variables;
+    * The mocks are defined and their methods overridden when necessary;
+    * The test subject is created and given the mocks;
```
[the remark this answers is not in the document]
Comment (Ullaakut) [reply content/code/testing.md:163]: I'm not sure what you mean by this comment, sorry 😅

File: content/include/function_example_tests.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                },
+                mockErr: mocks.GenericError,
+                checkErr: require.Error,
+            },
+        }
+
+        for _, test := range tests {
+            test := test
+            t.Run(test.name, func (t *testing.T) {
+                t.Parallel()
```
[the remark this answers is not in the document]
Comment (Ullaakut) [reply content/include/function_example_tests.md:38]: The only case where we should not use parallelization is when testing something that is not concurrency safe. There is zero downside to doing it otherwise, so we should always do it IMO 🤷

File: content/code/testing.md [hunk 7]
```diff
@@ -24,8 +31,37 @@ The best way to go about it is to follow the [official guidelines of the Go `tes
...
-The tests usually start with a subtest called `nominal case` which verifies that the tested component behaves as expected in a baseline situation, where no failures occur and no edge cases are handled.
+You might sometimes want to have multiple test functions for a single method, in which case the tests should be prefixed as defined above, and followed with the purpose of the test:
+
+```go
+func TestMyType_MethodDoesSomethingWhenX(t *testing.T) { ... }
+func TestMyType_MethodDoesHandlesXFailure(t *testing.T) { ... }
+// And so on.
+```
+
+When it comes to subtests, the names of individual subtests should be lowercased and concise. The tests usually start
```
[the remark this answers is not in the document]
Comment (Ullaakut) [reply content/code/testing.md:42]: WDYM?

File: content/code/testing.md (hunk 1 above)
Comment (Ullaakut) [reply content/code/testing.md:113]: I'm not sure what we should show as an example TBH 🤔 A function that generates one element of a given type deterministically and then another that generates a slice of it, like we have in Flow DPS? I feel like that might be misleading though, because they are implemented this way for the DPS but might be done differently somewhere else, so I'd rather have this part slightly open to interpretation 🤔 WDYT?

File: content/code/testing.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (emailtovamos) [reply content/code/testing.md:163]: Like test subject means a struct?

File: content/code/testing.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (emailtovamos) [reply content/code/testing.md:42]: I mean for example when to have 
```
func TestT_M() { ... }
func TestT_N() { ... }
```

as opposed to this: 

```
func TestT() {
t.Run()...
t.Run()...
}
```

Unless one should always follow one of the above.

File: content/code/testing.md (hunk 1 above)
Comment (emailtovamos) [reply content/code/testing.md:113]: Ok that makes sense.

File: content/code/testing.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ullaakut) [reply content/code/testing.md:42]: We would never do

```go
func TestT() {
	t.Run()...
	t.Run()...
}
```

Because that would mean we are testing the type itself. We always test methods, so in the case of your other example, we test two methods, and each method might have subtests, and/or even multiple tests for the same methods:

```go
func TestT_M() {
	t.Run()...
	t.Run()...
}

func TestT_MDoesSomething() {
	// ...
}

func TestT_N() {
	t.Run()...
	t.Run()...
}
```

[the remark this answers is not in the document]
Comment (Ullaakut) [reply content/code/testing.md:42]: As for when to use one function with tons of subtests to cover the whole method versus using multiple functions, I think there's no clear rule, but usually the point is that ideally we want to have one test function with subtests, and if one specific test case would be too different from the test, or require some very weird logic, we might want to separate it into a different test function. But once again, there's no clear rule for this :)

File: content/code/testing.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ullaakut) [reply content/code/testing.md:163]: If we are testing a method, the test subject is the struct on which we call the method, yes. If we're testing a function, there's no subject actually 🤔

WDYT of changing it to

```suggestion
    * If test is for a method, the method's struct (the subject of the test) is created and given the mocks;
```

[the remark this answers is not in the document]
Comment (emailtovamos) [reply content/code/testing.md:163]: Looks good 👍

**Review by AnNiran — APPROVED** [review]

File: content/code/testing.md [hunk 8]
```diff
@@ -24,8 +31,37 @@ The best way to go about it is to follow the [official guidelines of the Go `tes
 >   func TestT_M() { ... }
 > ```
 
-When it comes to subtests, the names of individual subtests should be lowercased and concise.
-The tests usually start with a subtest called `nominal case` which verifies that the tested component behaves as expected in a baseline situation, where no failures occur and no edge cases are handled.
+You might sometimes want to have multiple test functions for a single method, in which case the tests should be prefixed as defined above, and followed with the purpose of the test:
+
+```go
+func TestMyType_MethodDoesSomethingWhenX(t *testing.T) { ... }
+func TestMyType_MethodDoesHandlesXFailure(t *testing.T) { ... }
```
Comment (ricardomgoncalves) [inline content/code/testing.md:38]:
```suggestion
func TestMyType_MethodDoesHandleXFailure(t *testing.T) { ... }
```

Not really sure.

**Review by ricardomgoncalves — APPROVED** [review]
LGTM. Just a comment that I don't know if makes sense or not.

File: content/code/testing.md [hunk 9]
```diff
@@ -24,8 +31,37 @@ The best way to go about it is to follow the [official guidelines of the Go `tes
...
+With the following example:
+
+```go
+{!function_example.md!}
+```
+
+There are three possible paths through which this function can be traversed:
+
+* The nominal case, where `s.validate.Struct` and `s.index.HeightForBlock` return no errors and the function returns a
+  valid response.
```
Comment (Maelkum) [inline content/code/testing.md:56]: Is this newline accidental or to keep line shorter?

File: content/code/testing.md [hunk 10]
```diff
@@ -104,6 +150,20 @@ This can be very useful, as it enables creating a hierarchical structure within
...
+
+* At the top-level, the common testing variables that are used in multiple subtests.
+    * This can be a logger that discards logs, some test input that can be reused, expected return values shared between tests, and so on.
+* In each sub-test:
+    * The test is defined as parallelized;
+    * The inputs and outputs are defined in variables;
+    * The mocks are defined and their methods overridden when necessary;
+    * If test is for a method, the method's struct (the subject of the test) is created and given the mocks;
+    * The tested method is called on the subject, its return values are often stored in `got` and `err`: `got, err := mything.Do(input)`;
+    * Assertions are made on the returned values.
```
Comment (bogdanprodanj) [inline content/code/testing.md:165]: It's also nice to use `require.NoError(t, err)` on the `nominal case` tests. This will insure that test will fail before the assertions.

**Review by Maelkum — APPROVED** [review]
Looks good to me :)

File: content/code/testing.md (hunk 10 above)
Comment (Ullaakut) [reply content/code/testing.md:165]: It's what I implied by using `values` plural, but you're right maybe it's not clear enough. WDYT of

```suggestion
	* If the test should not error, `require.NoError` is called on the error return;
    * Assertions are made on the returned value.
```

File: content/code/testing.md (hunk 9 above)
Comment (Ullaakut) [reply content/code/testing.md:56]: It's accidental, I think it's my IDE's formatter that tries to keep lines below 80 characters, nice catch :D 

```suggestion
* The nominal case, where `s.validate.Struct` and `s.index.HeightForBlock` return no errors and the function returns a valid response.
```
